tmail-backend/jmap/extensions-es-v7/src/main/java/com/linagora/tmail/james/jmap/ESEmailAddressContactSearchEngine.java

```diff
@@ -79,22 +79,48 @@ public ESEmailAddressContactSearchEngine(ReactorElasticSearchClient client, Elas
 
     @Override
     public Publisher<EmailAddressContact> update(AccountId accountId, MailAddress mailAddress, ContactFields updatedFields) {
-        throw new NotImplementedException("Not implemented yet!");
+        if (mailAddress.equals(updatedFields.address())) {
+            EmailAddressContact updatedEmailAddressContact = EmailAddressContact.of(updatedFields);
+            return Mono.fromCallable(() -> mapper.writeValueAsString(new UserContactDocument(accountId, updatedEmailAddressContact)))
+                .flatMap(content -> userContactIndexer.update(
+                    ImmutableList.of(new UpdatedRepresentation(computeUserContactDocumentId(accountId, mailAddress), content)),
+                    RoutingKey.fromString(mailAddress.asString())))
+                .thenReturn(updatedEmailAddressContact);
+        } else {
+            return Mono.from(delete(accountId, mailAddress))
+                .then(Mono.from(index(accountId, updatedFields)));
```

Comment:
I think it would be more optimized if these two requests are done parallelly (the delete request not strictly be done first). Maybe something like this ?:
```java
            Mono<Void> deleteMono = Mono.from(delete(accountId, mailAddress));
            Mono<EmailAddressContact> indexMono = Mono.from(index(accountId, updatedFields));
            return Mono.zip(deleteMono, indexMono)
                .map(Tuple2::getT2);
```

rf: https://stackoverflow.com/questions/60850570/execute-three-mono-in-parallel-as-soon-they-are-created-wait-for-all-to-finish

An better solution maybe even just need to wait index be finished and don't care about the delete request finishment?

Reply:
Hmmm I'm not sure that it's really necessary TBH... other opinions on this?